internal/users/manager.go

```diff
@@ -237,6 +328,59 @@ func (m *Manager) UpdateUser(u types.UserInfo) (err error) {
 	return nil
 }
 
+func (m *Manager) getOldUserInfoFromDB(name string) (oldUserInfo *types.UserInfo, err error) {
+	// Check if the user already exists in the database.
+	u, err := m.db.UserByName(name)
+	if errors.Is(err, db.NoDataFoundError{}) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get user %q: %w", name, err)
+	}
+
+	userGroups, err := m.db.UserGroups(u.UID)
+	if err != nil && !errors.Is(err, db.NoDataFoundError{}) {
+		return nil, fmt.Errorf("could not get users groups for %q: %w", name, err)
+	}
+
+	oldLocalGroups, err := m.db.UserLocalGroups(u.UID)
+	if err != nil && !errors.Is(err, db.NoDataFoundError{}) {
+		return nil, err
+	}
```

Comment:
We perform multiple operations on the database here without taking a lock, that's prone to a race. I think if we use a singe database transaction instead, it should be fine.

Reply:
I'll draft something

Reply:
So, I did not check the change, but indeed that is potentially racy (by design) on the first call, since we want to just try and see, while it is not once we set the `userManagementMu` lock.

The `checkEqualUserExists` uses a quite clear pattern:
  - Do not lock anything and try for the best: a positive result is safe to use, so we do not want any locking
  - A negative result is actually a "maybe false" result, thus we lock and check with that state.

Now, this is fine since we do not have other functions acting on users at this level IIRC.

Reply:
Ok, saw the new commit: One single transaction is fine too, it increases the possibility of getting a result on racy conditions, but it's not changing the situation, so I am fine we go with that too.

Reply:
>  for example between the calls to m.db.UserGroups and m.db.UserLocalGroups the database entries are changed, we can end up with a user info that doesn't reflect any consistent state of the database.

Sure, sure... But it would likely not match the requested user then, or if it does, it means there are other requests coming or arrived that changed it, but I am fine with having it more consistent in a single transaction, so thanks for handling it